File: 백준/Silver/1912. 연속합/연속합.py
import sys
input = sys.stdin.readline

n = int(input())
n_list = list(map(int, input().split()))
dp = [n_list[0]]

for i in range(1, n):
    # 지금까지 누적된 최대값과 자기자신을 비교
    if dp[i - 1] + n_list[i] < n_list[i]:
        dp.append(n_list[i])
    else:
        dp.append(dp[i - 1] + n_list[i])
print(max(dp))

File: 백준/Silver/1912. 연속합/연속합.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
       BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        String[] input = br.readLine().split(" ");

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = Integer.parseInt(input[i]);
        }

        int[] d = new int[n];

        // d[i] = max(d[i-1] + a[i], a[i])
        for (int i = 0; i < n; i++) {
            d[i] = a[i];
            if (i > 0 && d[i] < d[i-1] + a[i]) {
                d[i] = d[i-1] + a[i];
            }
        }

        // 최댓값
        int ans = d[0];
        for (int i = 0; i < n; i++) {
            if (ans < d[i]) {
                ans = d[i];
            }
        }
        System.out.print(ans);
    }
}